### PR TITLE
Web-client helper methods

### PR DESCRIPTION
**Pull request by sisou**
## What's in this pull request?

Adds various helper methods to web-client structs that I found are used in our frontend-code, now that I am converting more of the apps.

- `.equals()` and `.compare()` methods for some primitives
- Contract creation/interaction data decoding methods

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

### CODE REVIEW

File: web-client/src/common/address.rs [hunk 1]
```diff
@@ -91,6 +91,16 @@ impl Address {
     pub fn serialize(&self) -> Vec<u8> {
         self.inner.serialize_to_vec()
     }
+
+    #[cfg(feature = "primitives")]
```
Comment (jsdanielh) [inline web-client/src/common/address.rs:97]: Could you add rustdoc?

File: web-client/src/common/address.rs [hunk 2]
```diff
@@ -91,6 +91,16 @@ impl Address {
...
+    pub fn equals(&self, other: &Address) -> bool {
+        self.inner == other.inner
+    }
+
+    #[cfg(feature = "primitives")]
```
Comment (jsdanielh) [inline web-client/src/common/address.rs:106]: Could you add rustdoc?

File: web-client/src/common/hashed_time_locked_contract.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+use crate::common::transaction::{
+    PlainHtlcData, PlainHtlcEarlyResolveProof, PlainHtlcRegularTransferProof,
+    PlainHtlcTimeoutResolveProof, PlainTransactionProof, PlainTransactionRecipientData,
+};
+#[cfg(feature = "primitives")]
+use crate::common::transaction::{PlainTransactionProofType, PlainTransactionRecipientDataType};
+
+#[wasm_bindgen]
+pub struct HashedTimeLockedContract;
```
Comment (jsdanielh) [inline web-client/src/common/hashed_time_locked_contract.rs:17]: Could you add rustdoc?

File: web-client/src/common/hashed_time_locked_contract.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+#[cfg(feature = "primitives")]
+#[wasm_bindgen]
+impl HashedTimeLockedContract {
+    #[wasm_bindgen(js_name = dataToPlain)]
+    pub fn data_to_plain(data: &[u8]) -> Result<PlainTransactionRecipientDataType, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/hashed_time_locked_contract.rs:24]: Could you add rustdoc?

File: web-client/src/common/hashed_time_locked_contract.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        let plain = HashedTimeLockedContract::parse_data(data)?;
+        Ok(serde_wasm_bindgen::to_value(&plain)?.into())
+    }
+
+    #[wasm_bindgen(js_name = proofToPlain)]
+    pub fn proof_to_plain(proof: &[u8]) -> Result<PlainTransactionProofType, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/hashed_time_locked_contract.rs:31]: Could you add rustdoc?

File: web-client/src/common/hashed_time_locked_contract.rs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        let plain = HashedTimeLockedContract::parse_proof(proof)?;
+        Ok(serde_wasm_bindgen::to_value(&plain)?.into())
+    }
+}
+
+impl HashedTimeLockedContract {
+    pub fn parse_data(bytes: &[u8]) -> Result<PlainTransactionRecipientData, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/hashed_time_locked_contract.rs:38]: Could you add rustdoc?

File: web-client/src/common/hashed_time_locked_contract.rs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+                AnyHash::Sha256(_) => "sha256".to_string(),
+                AnyHash::Sha512(_) => "sha512".to_string(),
+            },
+            hash_root: data.hash_root.to_hex(),
+            hash_count: data.hash_count,
+            timeout: data.timeout,
+        }))
+    }
+
+    pub fn parse_proof(bytes: &[u8]) -> Result<PlainTransactionProof, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/hashed_time_locked_contract.rs:56]: Could you add rustdoc?

File: web-client/src/common/signature_proof.rs [hunk 8]
```diff
@@ -175,16 +180,23 @@ impl SignatureProof {
     pub fn serialize(&self) -> Vec<u8> {
         self.inner.serialize_to_vec()
     }
+
+    pub fn to_plain(&self) -> Result<PlainTransactionProofType, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/signature_proof.rs:186]: Could you add rustdoc?

File: web-client/src/common/signature_proof.rs [hunk 9]
```diff
@@ -240,6 +252,33 @@ impl SignatureProof {
     }
 }
 
+impl SignatureProof {
+    pub fn to_plain_transaction_proof(&self) -> PlainTransactionProof {
```
Comment (jsdanielh) [inline web-client/src/common/signature_proof.rs:258]: Could you add rustdoc?

File: web-client/src/primitives/public_key.rs [hunk 10]
```diff
@@ -24,6 +24,16 @@ impl PublicKey {
 
 #[wasm_bindgen]
 impl PublicKey {
+    #[wasm_bindgen(getter = SIZE)]
+    pub fn size() -> usize {
```
Comment (jsdanielh) [inline web-client/src/primitives/public_key.rs:28]: Could you add rustdoc?

File: web-client/src/primitives/public_key.rs [hunk 11]
```diff
@@ -24,6 +24,16 @@ impl PublicKey {
...
+        nimiq_keys::Ed25519PublicKey::SIZE
+    }
+
+    #[wasm_bindgen(getter = serializedSize)]
+    pub fn serialized_size(&self) -> usize {
```
Comment (jsdanielh) [inline web-client/src/primitives/public_key.rs:33]: Could you add rustdoc?

File: web-client/src/primitives/public_key.rs [hunk 12]
```diff
@@ -101,6 +111,14 @@ impl PublicKey {
     pub fn to_address(&self) -> Address {
         Address::from(nimiq_keys::Address::from(&self.inner))
     }
+
+    pub fn equals(&self, other: &PublicKey) -> bool {
```
Comment (jsdanielh) [inline web-client/src/primitives/public_key.rs:116]: Could you add rustdoc?

File: web-client/src/primitives/public_key.rs [hunk 13]
```diff
@@ -101,6 +111,14 @@ impl PublicKey {
...
+        self.inner == other.inner
+    }
+
+    pub fn compare(&self, other: &PublicKey) -> i32 {
```
Comment (jsdanielh) [inline web-client/src/primitives/public_key.rs:124]: Could you add rustdoc?

File: web-client/src/primitives/private_key.rs [hunk 14]
```diff
@@ -70,6 +70,10 @@ impl PrivateKey {
     pub fn to_hex(&self) -> String {
         self.inner.to_hex()
     }
+
+    pub fn equals(&self, other: &PrivateKey) -> bool {
```
Comment (jsdanielh) [inline web-client/src/primitives/private_key.rs:75]: Could you add rustdoc?

File: web-client/src/primitives/es256_public_key.rs [hunk 15]
```diff
@@ -118,6 +115,14 @@ impl ES256PublicKey {
     pub fn to_address(&self) -> Address {
         Address::from(nimiq_keys::Address::from(&self.inner))
     }
+
+    pub fn equals(&self, other: &ES256PublicKey) -> bool {
```
Comment (jsdanielh) [inline web-client/src/primitives/es256_public_key.rs:120]: Could you add rustdoc?

File: web-client/src/primitives/es256_public_key.rs [hunk 16]
```diff
@@ -118,6 +115,14 @@ impl ES256PublicKey {
...
+        self.inner == other.inner
+    }
+
+    pub fn compare(&self, other: &ES256PublicKey) -> i32 {
```
Comment (jsdanielh) [inline web-client/src/primitives/es256_public_key.rs:128]: Could you add rustdoc?

File: web-client/src/common/transaction.rs [hunk 17]
```diff
@@ -820,24 +610,24 @@ pub struct PlainHtlcData {
 #[derive(Clone, serde::Serialize, serde::Deserialize, Tsify)]
 #[serde(rename_all = "camelCase")]
 pub struct PlainCreateValidatorData {
-    raw: String,
-    signing_key: String,
-    voting_key: String,
-    reward_address: String,
-    signal_data: Option<String>,
-    proof_of_knowledge: String,
+    pub raw: String,
```
Comment (jsdanielh) [inline web-client/src/common/transaction.rs:613]: Could you add rustdoc?

File: web-client/src/common/transaction.rs [hunk 18]
```diff
@@ -820,24 +610,24 @@ pub struct PlainHtlcData {
...
 #[derive(Clone, serde::Serialize, serde::Deserialize, Tsify)]
 #[serde(rename_all = "camelCase")]
 pub struct PlainUpdateValidatorData {
-    raw: String,
-    new_signing_key: Option<String>,
-    new_voting_key: Option<String>,
-    new_reward_address: Option<String>,
-    new_signal_data: Option<Option<String>>,
-    new_proof_of_knowledge: Option<String>,
+    pub raw: String,
```
Comment (jsdanielh) [inline web-client/src/common/transaction.rs:625]: Could you add rustdoc?

File: web-client/src/common/vesting_contract.rs [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+#[wasm_bindgen]
+pub struct VestingContract;
+
+#[cfg(feature = "primitives")]
+#[wasm_bindgen]
+impl VestingContract {
+    #[wasm_bindgen(js_name = dataToPlain)]
+    pub fn data_to_plain(data: &[u8]) -> Result<PlainTransactionRecipientDataType, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/vesting_contract.rs:21]: Could you add rustdoc?

File: web-client/src/common/vesting_contract.rs [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+        let plain = VestingContract::parse_data(data)?;
+        Ok(serde_wasm_bindgen::to_value(&plain)?.into())
+    }
+
+    #[wasm_bindgen(js_name = proofToPlain)]
+    pub fn proof_to_plain(proof: &[u8]) -> Result<PlainTransactionProofType, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/vesting_contract.rs:28]: Could you add rustdoc?

File: web-client/src/common/vesting_contract.rs [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+        let plain = VestingContract::parse_proof(proof)?;
+        Ok(serde_wasm_bindgen::to_value(&plain)?.into())
+    }
+}
+
+impl VestingContract {
+    pub fn parse_data(bytes: &[u8]) -> Result<PlainTransactionRecipientData, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/vesting_contract.rs:35]: Could you add rustdoc?

File: web-client/src/common/vesting_contract.rs [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+        Ok(PlainTransactionRecipientData::Vesting(PlainVestingData {
+            raw: hex::encode(bytes),
+            owner: data.owner.to_user_friendly_address(),
+            start_time: data.start_time,
+            step_amount: data.step_amount.into(),
+            time_step: data.time_step,
+        }))
+    }
+
+    pub fn parse_proof(bytes: &[u8]) -> Result<PlainTransactionProof, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/vesting_contract.rs:47]: Could you add rustdoc?

File: web-client/src/common/vesting_contract.rs [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+
+#[cfg(feature = "primitives")]
+use crate::common::transaction::{PlainTransactionProofType, PlainTransactionRecipientDataType};
+use crate::common::{
+    signature_proof::SignatureProof,
+    transaction::{PlainTransactionProof, PlainTransactionRecipientData, PlainVestingData},
+};
+
+#[wasm_bindgen]
+pub struct VestingContract;
```
Comment (jsdanielh) [inline web-client/src/common/vesting_contract.rs:14]: Could you add rustdoc?

File: web-client/src/common/staking_contract.rs [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+    transaction::{
+        PlainAddStakeData, PlainCreateStakerData, PlainCreateValidatorData, PlainRawData,
+        PlainRetireStakeData, PlainSetActiveStakeData, PlainTransactionProof,
+        PlainTransactionRecipientData, PlainUpdateStakerData, PlainUpdateValidatorData,
+        PlainValidatorData,
+    },
+};
+
+#[wasm_bindgen]
+pub struct StakingContract;
```
Comment (jsdanielh) [inline web-client/src/common/staking_contract.rs:19]: Could you add rustdoc?

File: web-client/src/common/staking_contract.rs [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+
+#[cfg(feature = "primitives")]
+#[wasm_bindgen]
+impl StakingContract {
+    #[wasm_bindgen(js_name = dataToPlain)]
+    pub fn data_to_plain(data: &[u8]) -> Result<PlainTransactionRecipientDataType, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/staking_contract.rs:26]: Could you add rustdoc?

File: web-client/src/common/staking_contract.rs [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+        let plain = StakingContract::parse_data(data)?;
+        Ok(serde_wasm_bindgen::to_value(&plain)?.into())
+    }
+
+    #[wasm_bindgen(js_name = proofToPlain)]
+    pub fn proof_to_plain(proof: &[u8]) -> Result<PlainTransactionProofType, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/staking_contract.rs:33]: Could you add rustdoc?

File: web-client/src/common/staking_contract.rs [hunk 27]
```diff
@@ -0,0 +1,134 @@
...
+        let plain = StakingContract::parse_proof(proof)?;
+        Ok(serde_wasm_bindgen::to_value(&plain)?.into())
+    }
+}
+
+impl StakingContract {
+    pub fn parse_data(bytes: &[u8]) -> Result<PlainTransactionRecipientData, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/staking_contract.rs:40]: Could you add rustdoc?

File: web-client/src/common/staking_contract.rs [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+                retire_stake,
+                proof: _proof,
+            } => PlainTransactionRecipientData::RetireStake(PlainRetireStakeData {
+                raw: hex::encode(bytes),
+                retire_stake: retire_stake.into(),
+            }),
+        })
+    }
+
+    pub fn parse_proof(bytes: &[u8]) -> Result<PlainTransactionProof, JsError> {
```
Comment (jsdanielh) [inline web-client/src/common/staking_contract.rs:133]: Could you add rustdoc?

**Comment by sisou** [comment]
@jsdanielh I rebased on latest `albatross` and added rustdocs to all WASM-exposed methods. I resolved those comments above. The remaining ones are methods only used internally between structs, not exposed to userland.

File: web-client/src/primitives/public_key.rs (hunk 10 above)
Comment (sisou) [reply web-client/src/primitives/public_key.rs:28]: Self-documenting.

File: web-client/src/primitives/public_key.rs (hunk 11 above)
Comment (sisou) [reply web-client/src/primitives/public_key.rs:33]: Self-documenting.

**Review by jsdanielh — APPROVED** [review]

**Pull request merged**